daemon: support alternative runtimes MVP

corhere opened the pull request:
- Closes #43736 
- Closes #42986

Contrary to popular belief, the OCI Runtime specification does not specify the command-line API for runtimes. Looking at containerd's architecture from the lens of the OCI Runtime spec, the _shim_ is the OCI Runtime and runC is "just" an implementation detail of the `io.containerd.runc.v2` runtime. When one configures a non-default runtime in Docker, what they're really doing is instructing Docker to create containers using the `io.containerd.runc.v2` runtime with a configuration option telling the runtime that the runC binary is at some non-default path. Consequently, only OCI runtimes which are compatible with the `io.containerd.runc.v2` shim, such as crun, can be used in this manner. Other OCI runtimes, including kata-containers v2, come with their own containerd shim and are not compatible with `io.containerd.runc.v2`. As Docker has not historically provided a way to select a non-default runtime which requires its own shim, runtimes such as kata-containers v2 could not be used with Docker.

Allow other ContainerD shims to be used with Docker; no daemon configuration required. If the daemon is instructed to create a container with a runtime name which does not match any of the configured or stock runtimes, it passes the name along to ContainerD verbatim. A user can start a container with the kata-containers runtime, for example, simply by calling

    docker run --runtime io.containerd.kata.v2

Runtime names which ContainerD would interpret as a path to an arbitrary binary are disallowed. While handy for development and testing it is not strictly necessary and would allow anyone with Engine API access to trivially execute any binary on the host as root, so we have decided it would be safest for our users if it was not allowed.

It is not yet possible to set an alternative ContainerD shim as the default runtime; it can only be configured per-container. Doing so properly, with validation and configuration reload handling, is quite the nontrivial undertaking with the current state of the daemon configuration logic. That will be addressed in a subsequent PR.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**
Unit and integration tests cover the new functionality.

To manually verify:
- Install an alternative runtime with a shim that implements Containerd Runtime v2 on `$PATH`
- `docker run --rm --runtime io.containerd.my-runtime.v1 alpine ls`

The `io.containerd.runc.v2` shim is aliased to `io.containerd.runc-alias.v2` in the development container image as a convenience for manual testing.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

- Alternative OCI runtimes compatible with the [Containerd Runtime v2](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md) API which are installed on the system can now be used to start containers. These runtimes are not listed in `docker info` and cannot be set as the default runtime at this time.

**- A picture of a cute animal (not mandatory but encouraged)**

cpuguy83 commented:
How about an integration test?
Could possibly just copy the runc shim to a different name (and path, should test both cases).

thaJeztah commented:
Let me post my comment from the other PR here for consideration https://github.com/moby/moby/pull/43736#discussion_r919738409

> From a UX perspective, the `--runtime` option was meant to be the more "convenient" option, where the user can use `docker info` to see which runtimes are available/configured on the daemon, and use those as option (`--runtime=runc`, `--runtime=kata`), and the daemon-configuratioin can contain additional config options (IIRC, this could even be multiple versions of `runc`, e.g.).
> 
> > No daemon.json configuration would be required to use alternative runtimes if the daemon was to interpret --runtime values
> 
> I know that access to the API already is equivalent to `root`, but if the `--runtime` value passed can be (as is being discussed here) either a name (`io.containerd.runc.v2`) or a path to a binary, that could potentially open up some "interesting" things (`--runtime=/bin/rm`, `--runtime=/some/arbitrary/script.sh`)

corhere commented:
> From a UX perspective, the `--runtime` option was meant to be the more "convenient" option, where the user can use `docker info` to see which runtimes are available/configured on the daemon, and use those as option (`--runtime=runc`, `--runtime=kata`), and the daemon-configuratioin can contain additional config options (IIRC, this could even be multiple versions of `runc`, e.g.).

And from a perspective of the dockerd<->containerd interface, the `--runtime` option looks a lot like picking a "configuration profile" which specifies some of the container level shim configuration the daemon should pass to the `io.containerd.runc.v2` [ContainerD Runtime](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#runtime-v2) on container creation. Extending this "configuration profile" concept to support defining a profile in daemon.json which specifies an arbitrary ContainerD Runtime with corresponding container level shim configuration is not very feasible with the current state of the ecosystem, at least not [with a UX that mortals could use](https://xkcd.com/378/) as the container level shim configuration [is passed to containerd inside a `protobuf.Any` message](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#container-level-shim-configuration). So I am making no attempt to support such extended profiles in this PR, though I have been mindful to not add further complications to building it out in the future. Without support for container level shim configuration, a configuration profile with an arbitrary ContainerD Runtime is little more than an alias.

One of my motivations for wanting the daemon to support using ContainerD Runtimes without needing to first explicitly inform the daemon of their existence is to remove a significant hurdle to packaging, installing and using alternative runtimes with Docker. Distributors could package runtimes purely as ContainerD Runtimes with no Docker awareness, and users would still be able to use the runtimes with Docker immediately after installation. And the daemon could dynamically discover which runtimes are available so they could be listed in `docker info` alongside the explicit runtime profiles. The ContainerD Runtime v2 spec [defines a mapping from runtime name to shim binary name](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#binary-naming). As this mapping is (more or less) invertible, installed ContainerD Runtimes could be discovered by searching `$PATH` for executables matching `^containerd-shim-([^-]+)-(v\d+)$` and mapping the file name to its runtime name `io.containerd.\1.\2`. (Ideally, containerd should be responsible for performing the search as its `$PATH` environment variable could differ from dockerd's.)

> I know that access to the API already is equivalent to `root`, but if the `--runtime` value passed can be (as is being discussed here) either a name (`io.containerd.runc.v2`) or a path to a binary, that could potentially open up some "interesting" things (`--runtime=/bin/rm`, `--runtime=/some/arbitrary/script.sh`)

As they say, with great power comes great responsibility. Containerd only accepts absolute paths, not relative, so it should hopefully be a red flag if someone instructs you to specify a runtime of `/some/arbitrary/script.sh`. If there is demand for it, we could add daemon config options for the more paranoid admins such as restricting the permissible `--runtime` values with an allowlist or disallowing runtimes-as-paths.

corhere commented on daemon/runtime_unix.go:
```diff
@@ -134,3 +137,35 @@ func (daemon *Daemon) getRuntime(name string) (*types.Runtime, error) {
 	return rt, nil
 }
+
+// isValidC8dRuntimeName tests whether name
+// is a well-formed ContainerD runtime name,
+// with additional Docker-specific restrictions.
+//
+// A runtime name containing slash characters is interpreted by ContainerD
+// as the path to a runtime binary.
+// If we allowed this, anyone with Engine API access could get ContainerD
+// to execute an arbitrary binary as root.
+// Although Engine API access is already equivalent to root on the host,
+// the runtime name has not historically been a vector
+// to run arbitrary code as root
+// so users are not expecting it to become one.
+//
+// This restriction is not configurable.
+// There are viable workarounds for legitimate use cases:
+// administrators and runtime developers can make runtimes available
+// for use with Docker by installing them onto $PATH
+// following the [binary naming convention] for ContainerD Runtime v2.
```
I'm giving [semantic linefeeds](https://rhodesmill.org/brandon/2012/one-sentence-per-line/) (https://sembr.org/) a try. I'm interested to hear your thoughts and opinions on it.

thaJeztah replied:
Hm... honestly; not a fan. I find it hard to read with the (for the reader) random line breaks.

I can see the advantage for _writing_ it (as well as in some cases to _update_), but (assuming the "happy path"), updates should be a lot less frequent than "reading". And while Markdown (probably as well when rendered as HTML on pkg.go.dev) would render it similar, I'm more often reading it as "source" than on the website 🤔

corhere replied:
I first learned about semantic linefeeds from [go.dev/doc/comment](https://go.dev/doc/comment#:~:text=The%20beginning%20of,when%20printing%20it.). It probably is overkill for unexported functions, and maybe even all commentary outside of packages intended for external users like `client` and `pkg/*` where the primary readers of the comments will likely be doing so from pkg.go.dev.

cpuguy83 commented on Dockerfile:
```diff
@@ -375,6 +375,7 @@ ENV DOCKER_BUILDTAGS="${DOCKER_BUILDTAGS}"
 WORKDIR /go/src/github.com/docker/docker
 VOLUME /var/lib/docker
 VOLUME /home/unprivilegeduser/.local/share/docker
+RUN ln -s /usr/local/bin/containerd-shim-runc-v2 /usr/local/bin/containerd-shim-runc-alias-v2
```
Can you move this to the test itself?

corhere replied:
I find it useful to have the test harness around for interactive dev/test use under `make shell`

cpuguy83 replied:
But then you can't run the test without running this.

cpuguy83 replied:
https://github.com/cpuguy83/docker-testing

Maybe the test can just check if it already exists and link it if not?

cpuguy83 commented on daemon/runtime_unix.go:
```diff
@@ -134,3 +137,35 @@ func (daemon *Daemon) getRuntime(name string) (*types.Runtime, error) {
+	// it would resolve to a binary named "containerd-shim---".
+	// https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/manager.go#L297-L317
+	// https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/shim/util.go#L83-L93
+	return !strings.ContainsRune(name, '/') && strings.ContainsRune(name, '.')
```
`filepath.Separator`?

corhere replied:
[ContainerD does a substring check against a literal `/` character.](https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/manager.go#L307-L310) I could add a `!filepath.IsAbs(name)` clause to cover the absolute-path-on-Windows case.

corhere committed:
daemon: support other containerd runtimes (MVP)

Contrary to popular belief, the OCI Runtime specification does not
specify the command-line API for runtimes. Looking at containerd's
architecture from the lens of the OCI Runtime spec, the _shim_ is the
OCI Runtime and runC is "just" an implementation detail of the
io.containerd.runc.v2 runtime. When one configures a non-default runtime
in Docker, what they're really doing is instructing Docker to create
containers using the io.containerd.runc.v2 runtime with a configuration
option telling the runtime that the runC binary is at some non-default
path. Consequently, only OCI runtimes which are compatible with the
io.containerd.runc.v2 shim, such as crun, can be used in this manner.
Other OCI runtimes, including kata-containers v2, come with their own
containerd shim and are not compatible with io.containerd.runc.v2.
As Docker has not historically provided a way to select a non-default
runtime which requires its own shim, runtimes such as kata-containers v2
could not be used with Docker.

Allow other containerd shims to be used with Docker; no daemon
configuration required. If the daemon is instructed to create a
container with a runtime name which does not match any of the configured
or stock runtimes, it passes the name along to containerd verbatim. A
user can start a container with the kata-containers runtime, for
example, simply by calling

    docker run --runtime io.containerd.kata.v2

Runtime names which containerd would interpret as a path to an arbitrary
binary are disallowed. While handy for development and testing it is not
strictly necessary and would allow anyone with Engine API access to
trivially execute any binary on the host as root, so we have decided it
would be safest for our users if it was not allowed.

It is not yet possible to set an alternative containerd shim as the
default runtime; it can only be configured per-container.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere commented on testutil/daemon/daemon.go:
```diff
@@ -334,9 +335,9 @@ func (d *Daemon) StartWithLogFile(out *os.File, providedArgs ...string) error {
 			"-u", d.rootlessUser.Username,
-			"-E", "XDG_RUNTIME_DIR="+d.rootlessXDGRuntimeDir,
-			"-E", "HOME="+d.rootlessUser.HomeDir,
-			"-E", "PATH="+os.Getenv("PATH"),
+			"--preserve-env",
+			"XDG_RUNTIME_DIR="+d.rootlessXDGRuntimeDir,
+			"HOME="+d.rootlessUser.HomeDir,
```
Note for reviewers: the `sudo -E` flag takes no arguments. `sudo -E HOME=/home/rootlessuser` is parsed as two separate flags: preserve all environment variables, and set the `HOME` variable. The extra `-E` flags were redundant. The only functional change I have made is to stop overriding the `PATH` variable.

thaJeztah replied:
What was the reason for removing PATH from the list? Depending on the environment, PATH may be a protected env-var, and therefore reset to its default (which can lead to fun situations; see https://github.com/moby/term/pull/21#issuecomment-720093769)

corhere replied:
I was hoping to pass through `PATH` from the `os/exec.Cmd.Args` so the path override in my integration test (`daemon.WithEnvVars`) would Just Work in rootless mode. Guess I'll need to find a different way to accomplish that.

thaJeztah replied:
I guess that would still be possible 🤔 as in; when it would run `sudo -E PATH=/foo:/bar PATH=/bar/baz`, the second `PATH` would override the former (default "inherited") path, correct?

thaJeztah commented on Dockerfile:
```diff
@@ -375,6 +375,7 @@ ENV DOCKER_BUILDTAGS="${DOCKER_BUILDTAGS}"
 WORKDIR /go/src/github.com/docker/docker
 VOLUME /var/lib/docker
 VOLUME /home/unprivilegeduser/.local/share/docker
+RUN ln -s /usr/local/bin/containerd-shim-runc-v2 /usr/local/bin/containerd-shim-runc-alias-v2
```
I'm still a bit on the fence on this one (adding a layer just for the convenience of manually trying this construct).

Curious though; IIRC, you have an interest in using `crun` instead of `runc` - perhaps a weird idea, but would it make sense to add `crun` in the dev environment? This would also (if we want to) allow us to have CI run with `crun` instead of `runc`.

thaJeztah commented on daemon/runtime_unix.go:
```diff
@@ -134,3 +137,35 @@ func (daemon *Daemon) getRuntime(name string) (*types.Runtime, error) {
+	// it would resolve to a binary named "containerd-shim---".
+	// https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/manager.go#L297-L317
+	// https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/shim/util.go#L83-L93
+	return !filepath.IsAbs(name) && !strings.ContainsRune(name, '/') && strings.ContainsRune(name, '.')
```
This path would always be "local", for containerd, right? (i.e., no need to take cross-platform into account) otherwise we would need something like https://github.com/docker/cli/pull/1990


I think ideally;

- we should consume some function from containerd that performs the validation, so that we don't diverge
- perhaps create a tracking issue to contribute such a function in containerd (i.e. [based on this code](https://github.com/containerd/containerd/blob/11ded166c15f92450958078cd13c6d87131ec563/runtime/v2/manager.go#L297-L317)
- wondering if the containerd validation is (per the above) _too_ relaxed, i.e., does it want to allow `con....tai-----nerd-🤣🐳-shim...---` as name? (something like [`RestrictedNameChars`](https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/names/names.go#L6), possibly tweaked to not allow consecutive separators may be good?

thaJeztah commented on daemon/runtime_unix.go:
```diff
@@ -134,3 +137,35 @@ func (daemon *Daemon) getRuntime(name string) (*types.Runtime, error) {
 }
+
+// isValidC8dRuntimeName tests whether name
+// is a well-formed ContainerD runtime name,
```
note that the canonical name for containerd is all lowercase (`containerd`)

corhere replied:
🤦 I must've mixed it up with runC

thaJeztah replied:
I don't think there's an "official" statement, but afaik, it's all lowercase (but I know _some_ people use the capital D) - as we're using lowercase everywhere, best to stick with it 😅

thaJeztah reviewed (commented):
overall looks good; left some thoughts / ramblings

corhere commented on daemon/runtime_unix.go:
```diff
@@ -116,7 +117,10 @@ func (daemon *Daemon) rewriteRuntimePath(name, p string, args []string) (string,
 	rt := daemon.configStore.GetRuntime(name)
 	if rt == nil {
-		return nil, errdefs.InvalidParameter(errors.Errorf("runtime not found in config: %s", name))
+		if !isPermissibleC8dRuntimeName(name) {
+			return nil, errdefs.InvalidParameter(errors.Errorf("unknown or invalid runtime name: %s", name))
+		}
+		return &types.Runtime{Shim: &types.ShimConfig{Binary: name}}, nil
```
This PR is also laying some groundwork for the follow-up PR to enable a containerd runtime to be set as the default runtime. That function is not returning an error directly because the appropriate error to return is context-dependent. A different error message is more appropriate when validating the configured default runtime.

> Of course at that point it's debatable if it should also handle converting it to a `types.Runtime`, so something like;
> 
> ```go
> return getContainerdRuntime(name)
> ```

That hypothetical `getContainerdRuntime` function would be a near drop-in replacement for `(*Daemon).getRuntime`, having the same signature and everything. And it would almost seem to work, too. But it would be subtly broken to call it _anywhere_ other than `(*Daemon).getRuntime` as it would disregard any configured runtimes. I would prefer to have only one function which maps a name to a runtime so there is no possibility to choose wrong.

thaJeztah replied:
> But it would be subtly broken to call it anywhere other than (*Daemon).getRuntime as it would disregard any configured runtimes. 

Overall I'm not _too_ concerned (for a non-exported function, which more commonly come with some guidance). Was mostly considering it to be nice if it would return a nicely formatted error with the right type ("invalid name"), and generally `err` can be used as a "boolean" for validation.

Not a blocker though.

corhere commented on daemon/runtime_unix.go:
```diff
@@ -116,7 +117,10 @@ func (daemon *Daemon) rewriteRuntimePath(name, p string, args []string) (string,
+		return &types.Runtime{Shim: &types.ShimConfig{Binary: name}}, nil
 	}
 
 	if len(rt.Args) > 0 {
```
> small observation that all code below is _only_ applied to runtimes configured through `daemon.json`; for clarity we could consider reversing the `if rt == nil {` to `if rt != nil {` with those options below in it.

`rt == nil` is an error condition [so its flow is indented.](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow)
 
> That said, as I see it correctly;
> 
> * setting a runtime through `daemon.json` ~without specifying `Shim`,~ would result in `defaultV2ShimConfig()` to be applied. Whereas passing the same through ~`--runtime`~ `dockerd --add-runtime` would not. Which means that I cannot ~"just"~ specify `io.containerd.FOOBAR.v2` in `daemon.json` (as that would apply the defaultOptions, assuming it's a variant of `runc` ~shim~ CLI binary with the file name `io.containerd.FOOBAR.v2`)
> * `Args` can be passed through `daemon.json` (but obviously not through ~`--runtime`~ `--add-runtime`

FTFY. `runtimes.<name>.Shim` cannot be specified in `daemon.json` as that struct field is tagged `json:"-"`.

> Not sure what the best migration path is to allow "just" specifying a shim through daemon.json, and/or if some of those options should be marked deprecated and/or produce an error. (input/ideas welcome)

That's the neat part; you don't! "Just" specify the shim name as the runtime when creating the container. As noted in the PR description, I will be following up with a PR to allow `default-runtime` to be set to a shim name. No migration path is needed as `daemon.json` has exactly the same semantics as it always had. People can continue to define "runtimes" which use runC-compatible binaries with the `io.containerd.runc.v2` containerd runtime, same as they always have done. I see no need to deprecate that functionality, either.

corhere commented on daemon/daemon_unix.go:
```diff
@@ -707,8 +707,8 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.
 		hostConfig.Runtime = daemon.configStore.GetDefaultRuntimeName()
```
This _is_ mutating the config. I cringed when I first noticed it.

cpuguy83 commented:
What's up with the vendoring on this? I would not expect any changes to vendor/.

thaJeztah commented:
> What's up with the vendoring on this? I would not expect any changes to vendor/.

Looks like it's because of the `shim.Binaryname()` being used, and runc options in one of the tests

corhere commented:
`shim.BinaryName()` is the only reason for the vendoring explosion. The package with the runc options used in that test is also imported by existing code.

cpuguy83 commented:
Ah, I totally missed that shim import.

```
// BinaryName returns the shim binary name from the runtime name,
// empty string returns means runtime name is invalid
func BinaryName(runtime string) string {
	// runtime name should format like $prefix.name.version
	parts := strings.Split(runtime, ".")
	if len(parts) < 2 {
		return ""
	}

	return fmt.Sprintf(shimBinaryFormat, parts[len(parts)-2], parts[len(parts)-1])
}
```

Seems like we can copy this one.

corhere commented:
> Seems like we can copy this one.

That's what I did in the first revision! @thaJeztah wasn't happy about it: https://github.com/moby/moby/pull/43800#discussion_r926600198

thaJeztah commented:
> That's what I did in the first revision! @thaJeztah wasn't happy about it: https://github.com/moby/moby/pull/43800#discussion_r926600198

Yes, I guess there's pros/cons (no risk of diverging, but more code vendored).

- Is that a concern for you? Or "OK"? @cpuguy83 
- Will we be using more bits from the `shim` package going forward (with the containerd integration)?
- If not (and if there's a risk that the shim package also will cause more _indirect_ dependencies; we could consider looking if we can refactor some bits in containerd and to move such utilities to a smaller package (could be part of "config" validation in general).

Happy to hear your thoughts!

thaJeztah reviewed (approved):
LGTM
